lxd/apparmor/rsync.go

```diff
@@ -0,0 +1,221 @@
+//go:build linux && cgo && !agent
+
+package apparmor
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"text/template"
+
+	"github.com/lxc/lxd/lxd/sys"
+	"github.com/lxc/lxd/shared"
+	"github.com/lxc/lxd/shared/subprocess"
+)
+
+var rsyncProfileTpl = template.Must(template.New("rsyncProfile").Parse(`#include <tunables/global>
+profile "{{ .name }}" flags=(attach_disconnected,mediate_deleted) {
+  #include <abstractions/base>
+
+  capability dac_override,
+  capability dac_read_search,
+  capability mknod,
+  capability chown,
+  capability fsetid,
+  capability fowner,
+
+  unix, # TODO Perhaps a stricter restriction can be made here
+  
+  @{PROC}/@{pid}/cmdline r,
+  @{PROC}/@{pid}/cpuset r,
+  {{ .rootPath }}/{etc,lib,usr/lib}/os-release r,
+
+  {{ .logPath }}/*/netcat.log rw,
+
+  {{ .rootPath }}/run/{resolvconf,NetworkManager,systemd/resolve,connman,netconfig}/resolv.conf r,
+  {{ .rootPath }}/run/systemd/resolve/stub-resolv.conf r,
+
+{{range $index, $element := .allowedCmdPaths}}
+  {{$element}} mixr,
+{{- end }}
+
+  {{ .sourcePath }}/** r,
+  {{ .sourcePath }}/ r,
+
+
+{{- if .dstPath }}
+  {{ .dstPath }}/** rw,
+  {{ .dstPath }}/ rw,
```

Comment:
I have not tested but would have expected that it would also need `link` as in `rwl,`.

Reply:
I have tested it on my clusters. One cluster consists of 3 nodes and uses Ceph. The over is one LXD node, but also uses Ceph and dir as a storage.

To trigger LocalCopy I use:
```
lxc storage create test dir
lxc storage volume create test vl_test1
lxc storage volume snapshot test vl_test1
lxc storage volume restore test vl_test1 snap0
```

To trigger Send and Recv use:
```
lxc copy tc1 cl1:testCopy5
```
where `tc1` is a container and `cl1` is a remote cluster.

Reply:
Could you try copying a filesystem that contains hardlinks?